### ISSUE TITLE
Conflict with bom support in Gradle 5.2

### ISSUE THREAD

**Issue by yschimke** [gradle/gradle#8420]
Hit when upgrading https://github.com/yschimke/okurl

```
$ ./gradlew wrapper --gradle-version 5.2
$ ./gradlew test --info
```

```
> Task :compileKotlin FAILED
:compileKotlin (Thread[Execution worker for ':',5,main]) completed. Took 0.586 secs.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileKotlin'.
> Could not resolve all dependencies for configuration ':compileClasspath'.
   > Cannot choose between com.fasterxml.jackson:jackson-bom:2.9.8 variant compile and com.fasterxml.jackson:jackson-bom:2.9.8 variant platform-compile because they provide the same capability: com.fasterxml.jackson:jackson-bom:2.9.8

```

**Comment by melix** [comment]
There are two distinct issues here. First of all, the message isn't helpful, because even if you run:

```
./gradlew dependencyInsight --configuration compileClasspath --dependency jackson-bom
```
Gradle will fail the same way and not give you the offending path. This needs to be fixed.

So I debugged and identified the problem:

- you have a first level dependency of type `platform` on `jackson-bom`, which is the right thing to do
- however, there is also a dependency on `com.baulsupp:oksocial-output:4.25.0` which brings a transitive dependency to `jackson-bom` too. This one is a regular Maven dependency (not a platform one). Given that `jackson-bom` is a real platform and that there are no dependencies in here, this dependency is redundant and brings nothing, so it's an error of `oksocial-output`.

So what can we do here? First, `platform` is a way to accommodate this problem that when we see a Maven published component we need to make the difference between the "platform" and the "library" aspects of it. In 5.2, we made it an error to depend on two variants of components which provide the same capability, which is exactly what you have here: one is asking for the "platform" (your direct dependency), the other one is asking for the "library". However, in reality we're playing gods here, and the "things" published at the same coordinates (jackson-bom library and jackson-bom platform) are really two different _components_. So I think we could workaround the problem by assigning a different implicit capability to the platform component.

Note that it would fix your problem, but it wouldn't change the fact that you depend on 2 different things, it would just make it less visible.

A good way to fix, immediately, is to create a component metadata rule to fix the dependencies of `oksocial-output`:


```
dependencies {
   components.all {
      run {
         if (id.group == "com.baulsupp" && id.name == "oksocial-output") {
             allVariants {
                withDependencies {
                   removeAll { it.name == "jackson-bom" }
                }
             }
         }
     }
   }
}
```
